app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala

```diff
@@ -21,7 +21,7 @@ import java.util.UUID
 import uk.gov.hmrc.apisubscriptionfields.model.{Fields, SubscriptionIdentifier}
 
 object SubscriptionFields {
-  def apply(id: SubscriptionIdentifier, fieldsId: UUID, customFields: Fields) : SubscriptionFields = SubscriptionFields(id.encode(), fieldsId, customFields)
+  def apply(id: SubscriptionIdentifier, fieldsId: UUID, customFields: Fields) : SubscriptionFields = new SubscriptionFields(id.encode(), id.applicationId.value, fieldsId, customFields)
 }
 
-case class SubscriptionFields(id: String, fieldsId: UUID, customFields: Fields)
+case class SubscriptionFields(id: String, applicationId: String, fieldsId: UUID, customFields: Fields)
```

Comment:
The idea of adding `applicationId` as additional field and index in this mongo repository is a good idea.
This in fact allow you to find efficiently records by `applicationId`.

However, if you follow this approach, you will have repetitions in `id` and `applicationId`, as `id` contains the `applicationId`.

With this change, I think it is best to remove the `id` field, and introduce a new field called `apiIdentifier`.
The value of the field `apiIdentifier` could be `<apiContext>__<apiVersion>`.

In this mongo repository we would need:
- an index for the single field `applicationId`
- an index for the two fields `applicationId` and `apiIdentifier`

What do you think about this optimisation?

Obviously this change will need a some refactoring everywhere in the code, but I believe this is the way to go.

Reply:
The original `id` field is an encoded one of format `{application-id}___{api-context}___{api-version}` the encode process may add extra separator chars in first part of the above key. This means that the index on the above key can not be used for plain `applicationId` eg regex match using `^appId` (as indexed field may contain extra separator chars). So that's why we introduced the secondary `applicationId` index. My understanding is that we *both* indexes as we still need to do single record operations using `{application-id}___{api-context}___{api-version}` as well as bulk operations using `applicationId`

Reply:
At the moment the separator is `##`, not `__`.
Thus, the `id` field contains `{applicationId}___{apiContext}___{apiVersion}`.

I know we need the index on the field `applicationId`, as you need it when you fetch by `applicationId`.

But now the field `id` could be removed.
If you have the fields `applicationId` and `apiIdentifier` (where `apiIdentifier` is `{apiContext}___{apiVersion}`), you can introduce two mongo indexes:
1) one index on the single field `applicationId`
2) one index on the fields `applicationId` and `apiIdentifier`

When you fetch subscription fields using both the application id and the API (context and version), you need to use a selector on both `applicationId` and `apiIdentifier`.
This fetch operation will use the second index mentioned above.

We do not need to split `apiIdentifier` in two fields.
You can have both in the same field, using a separator (for example `__` or `##`).

Reply:
OK got it. Sounds like a good idea but a fair deal of work. Maybe a candidate for another (next) PR?

Reply:
If there is something new to add, I am happy for the work to be done in a new PR.
But since this comment is about code you committed in this PR, I think it is best to adjust it here in this PR, before to merge to master.